Bottom: 4cb90c0f4d2bc846557dd664d14cf0a5c9dc0c4a
Top:    98b4567fe79be47047097ec25fc452d11790955c
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-30 15:20:40 +0100

autoroute.c: Utilise new HID interface for debug drawing


---

diff --git a/src/autoroute.c b/src/autoroute.c
index f34451f..5c767ea 100644
--- a/src/autoroute.c
+++ b/src/autoroute.c
@@ -138,6 +138,7 @@ return dir;
 }
 
 #ifdef ROUTE_DEBUG
+HID *ddraw = NULL;
 static hidGC ar_gc = 0;
 #endif
 
@@ -1468,16 +1469,17 @@ showbox (BoxType b, Dimension thickness, int group)
   if (showboxen != -1 && showboxen != group)
     return;
 
-
-  gui->set_line_width (ar_gc, thickness);
-  gui->set_line_cap (ar_gc, Trace_Cap);
-  gui->set_color (ar_gc, SLayer->Color);
-
-  gui->draw_line (ar_gc, b.X1, b.Y1, b.X2, b.Y1);
-  gui->draw_line (ar_gc, b.X1, b.Y2, b.X2, b.Y2);
-  gui->draw_line (ar_gc, b.X1, b.Y1, b.X1, b.Y2);
-  gui->draw_line (ar_gc, b.X2, b.Y1, b.X2, b.Y2);
-  gui->use_mask (HID_FLUSH_DRAW_Q);
+  if (ddraw != NULL)
+    {
+      ddraw->set_line_width (ar_gc, thickness);
+      ddraw->set_line_cap (ar_gc, Trace_Cap);
+      ddraw->set_color (ar_gc, SLayer->Color);
+
+      ddraw->draw_line (ar_gc, b.X1, b.Y1, b.X2, b.Y1);
+      ddraw->draw_line (ar_gc, b.X1, b.Y2, b.X2, b.Y2);
+      ddraw->draw_line (ar_gc, b.X1, b.Y1, b.X1, b.Y2);
+      ddraw->draw_line (ar_gc, b.X2, b.Y1, b.X2, b.Y2);
+    }
 
 #if 1
   if (b.Y1 == b.Y2 || b.X1 == b.X2)
@@ -1522,23 +1524,26 @@ showedge (edge_t * e)
 {
   BoxType *b = (BoxType *) e->rb;
 
-  gui->set_line_cap (ar_gc, Trace_Cap);
-  gui->set_line_width (ar_gc, 1);
-  gui->set_color (ar_gc, Settings.MaskColor);
+  if (ddraw == NULL)
+    return;
+
+  ddraw->set_line_cap (ar_gc, Trace_Cap);
+  ddraw->set_line_width (ar_gc, 1);
+  ddraw->set_color (ar_gc, Settings.MaskColor);
 
   switch (e->expand_dir)
     {
     case NORTH:
-      gui->draw_line (ar_gc, b->X1, b->Y1, b->X2, b->Y1);
+      ddraw->draw_line (ar_gc, b->X1, b->Y1, b->X2, b->Y1);
       break;
     case SOUTH:
-      gui->draw_line (ar_gc, b->X1, b->Y2, b->X2, b->Y2);
+      ddraw->draw_line (ar_gc, b->X1, b->Y2, b->X2, b->Y2);
       break;
     case WEST:
-      gui->draw_line (ar_gc, b->X1, b->Y1, b->X1, b->Y2);
+      ddraw->draw_line (ar_gc, b->X1, b->Y1, b->X1, b->Y2);
       break;
     case EAST:
-      gui->draw_line (ar_gc, b->X2, b->Y1, b->X2, b->Y2);
+      ddraw->draw_line (ar_gc, b->X2, b->Y1, b->X2, b->Y2);
       break;
     default:
       break;
@@ -3665,6 +3670,11 @@ TracePath (routedata_t * rd, routebox_t * path, const routebox_t * target,
 
   if (TEST_FLAG (LIVEROUTEFLAG, PCB))
     Draw ();
+
+#ifdef ROUTE_DEBUG
+  if (ddraw != NULL)
+    ddraw->flush_debug_draw ();
+#endif
 }
 
 /* create a fake "edge" used to defer via site searching. */
@@ -5109,6 +5119,15 @@ AutoRoute (bool selected)
   total_wire_length = 0;
   total_via_count = 0;
 
+#ifdef ROUTE_DEBUG
+  ddraw = gui->request_debug_draw ();
+  if (ddraw != NULL)
+    {
+      ar_gc = ddraw->make_gc ();
+      ddraw->set_line_cap (ar_gc, Round_Cap);
+    }
+#endif
+
   for (i = 0; i < NUM_STYLES; i++)
     {
       if (PCB->RouteStyle[i].Thick == 0 ||
@@ -5271,6 +5290,11 @@ donerouting:
 	  r_search (rd->layergrouptree[i], &big, NULL, ripout_livedraw_obj_cb, NULL);
 	}
     }
+#ifdef ROUTE_DEBUG
+  if (ddraw != NULL)
+    ddraw->finish_debug_draw ();
+#endif
+
   if (changed)
     changed = IronDownAllUnfixedPaths (rd);
   Message ("Total added wire length = %f inches, %d vias added\n",
